packages/react-components/src/components/VideoGallery/videoGalleryUtils.ts

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { useRef } from 'react';
+import { smartDominantSpeakerParticipants } from '../../gallery';
+import { VideoGalleryParticipant, VideoGalleryRemoteParticipant } from '../../types';
+
+type LayoutResult = {
+  gridParticipants: VideoGalleryParticipant[];
+  horizontalGalleryParticipants: VideoGalleryParticipant[];
+};
+
+const DEFAULT_MAX_REMOTE_VIDEOSTREAMS = 4;
+
+const DEFAULT_MAX_AUDIO_DOMINANT_SPEAKERS = 6;
+
+/**
+ * @private
+ */
+export const useFloatingLocalVideoLayout = (props: {
+  remoteParticipants: VideoGalleryRemoteParticipant[];
+  dominantSpeakers?: string[];
+  maxRemoteVideoStreams?: number;
+  maxAudioDominantSpeakers?: number;
+  isScreenShareActive?: boolean;
+}): LayoutResult => {
+  const visibleVideoParticipants = useRef<VideoGalleryRemoteParticipant[]>([]);
+  const visibleAudioParticipants = useRef<VideoGalleryRemoteParticipant[]>([]);
+
+  const {
+    remoteParticipants,
+    dominantSpeakers,
+    maxRemoteVideoStreams = DEFAULT_MAX_REMOTE_VIDEOSTREAMS,
+    maxAudioDominantSpeakers = DEFAULT_MAX_AUDIO_DOMINANT_SPEAKERS,
+    isScreenShareActive = false
+  } = props;
+
+  visibleVideoParticipants.current = smartDominantSpeakerParticipants({
+    participants: remoteParticipants?.filter((p) => p.videoStream?.isAvailable) ?? [],
+    dominantSpeakers,
+    lastVisibleParticipants: visibleVideoParticipants.current,
+    maxDominantSpeakers: maxRemoteVideoStreams
+  }).slice(0, maxRemoteVideoStreams);
+
+  const visibleVideoParticipantsSet = new Set(visibleVideoParticipants.current.map((p) => p.userId));
+
+  /* @conditional-compile-remove(PSTN-calls) */ /* @conditional-compile-remove(one-to-n-calling) */
+  const callingParticipants = remoteParticipants.filter((p) => p.state === ('Connecting' || 'Ringing'));
+  /* @conditional-compile-remove(PSTN-calls) */ /* @conditional-compile-remove(one-to-n-calling) */
+  const visibleCallingParticipantsSet = new Set(callingParticipants.map((p) => p.userId));
+
+  visibleAudioParticipants.current = smartDominantSpeakerParticipants({
+    participants:
+      remoteParticipants?.filter(
+        (p) =>
+          !visibleVideoParticipantsSet.has(p.userId) &&
+          /* @conditional-compile-remove(PSTN-calls) */ /* @conditional-compile-remove(one-to-n-calling) */ !visibleCallingParticipantsSet.has(
+            p.userId
+          )
+      ) ?? [],
+    dominantSpeakers,
+    lastVisibleParticipants: visibleAudioParticipants.current,
+    maxDominantSpeakers: maxAudioDominantSpeakers
+  });
+
+  const getGridParticipants = (): VideoGalleryRemoteParticipant[] => {
```

Comment:
Let's memoize this so that this hook doesn't cause unnecessary renders in the component it's used :)

Reply:
You can use React.useCallback